### ISSUE TITLE
Some items in the status bar are not read by orca when we use the flat review mode to explore it.

### ISSUE THREAD

**Issue by jvesouza** [microsoft/vscode#94962]
Issue Type: <b>Bug</b>

1. Launch orca.
2. Open a file.
3. Use orca's command speaks the status bar.
4. Use orca's command moves flat review to the end position.

The content of the status bar is read by the orca but slightly different from what was read in step 3.
In my environment the two first itens, remote and branch name are not announced by orca.
The problems also happens when I use move flat review to the next item or move flat review to the previous item.
CC @joanmarie @isidorn 

VS Code version: Code - Insiders 1.45.0-insider (44c5185373d5f209cd6c0b5d29a216fb801da34e, 2020-04-10T18:15:17.029Z)
OS version: Linux x64 5.6.3-arch1-1

**Comment by isidorn** [comment]
@joanmarie commented: 
"When reading generic text objects like divs in browse mode (i.e. reading with arrows), Orca completely ignores the names of those text objects because (imho) generic text objects which are caret navigable and not focusable arguably shouldn't have names. I believe others share my view -- potentially including NVDA developers (though I cannot speak to that with certainty).

But when Orca's command to present the status bar is used, Orca examines the children and if those children have names, it happily accepts those names without question. Whether or not Orca should do that is another matter. But that's what Orca does at the present time. Perhaps that's what NVDA does as well?"

@jvesouza what is exactly the difference between what orca speaks and when you move? Is there some items missing, like which one for example?
Or is the order just different?

**Comment by jvesouza** [comment]
> 
> @jvesouza what is exactly the difference between what orca speaks and when you move? Is there some items missing, like which one for example?
> Or is the order just different?

Item 1 and item 2 are not read by orca.

**Comment by joanmarie** [comment]
So what appears to be happening is that Orca actually is speaking things -- or more accurately sending things to be spoken to speech-dispatcher. But Orca is diving deep down into the accessibility tree looking for items, and finding stuff.

Examples:
* For the div with "Remote" as its name, it finds a link with U+EB3A as its text and sends that off to be spoken (functional result: <crickets>)
* For the div with "Synchronize Changes" as its name, it finds a link with U+EA77 as its text and sends that off to be spoken (functional results: <crickets>)
* Etc.

The above I'll address in Orca so that the displayed text gets presented.

In addition, the order IS different:
* Screen Reader Optimized is the last child in the accessibility tree so it's presented last by Orca's statusbar command which presents things in child order.
* Screen Reader Optimized is visually right before the line and column count, so it's presented right before the line and column count in Orca's flat review mode which provides a spatially-based representation of what's on screen.

The ordering difference is IMHO not an Orca bug. Potentially this presentation is a feature of Orca + VSCode: The status bar is ordered based on perceived importance/relevance, but flat review can still be used to get the on-screen view (once I fix the unpresentable/meaningless-characters problem).

**Comment by isidorn** [comment]
@joanmarie yes the order of the elements in the status bar is not the same as how they are presented. This is an implementation bug which is tracked by this item https://github.com/microsoft/vscode/issues/76010

So I think you will tackle this one on the Orca side and the order we have the other issue so I think this one can be closed. Closing this and let me know if I should reopen. Thanks!

**Comment by joanmarie** [comment]
@jvesouza: Sorry for the delay (went down some flat review rabbit holes on the way to your fix). Mind pulling master and letting me know how things are working?

@isidorn: The one notable difference I'm seeing now in Orca's presentation is that when you read the status bar, Orca says "No problems". But when you flat review, Orca says what sounds like "0 0".

As I mentioned the other day, in flat review Orca digs down looking for presentable info. The way I fixed things like "Remote" (child link with U+EB3A as its full text) and "Synchronize Changes" (child link with U+EA77 as its full text) is to deem the links as having no end-user-presentable info and then have Orca ignore things with no end-user-presentable info. In contrast, the "No Problems" has a link whose text is "'\uea87 0 \uea6c 0'". The zeroes seem presentable. And filtering out any link with private-use-area symbols would cause certain share links to be ignored (imagine the Twitter symbol followed by "Tweet!")

So.... Any chance of making the link name "No Problems" (i.e. via `aria-label`)?

**Comment by jvesouza** [comment]
It was excellent! Thank you.

**Comment by isidorn** [comment]
@joanmarie thanks a lot for trying it out and fixing. I chekced and when there are 0 problems and 0 warning we nicely set the `aria-label` on that element to be "No Problems". I am not sure what else could we do, or you do not see the `aria-label` being set on that element?

Here's the full html of that element
```html
<div id="status.problems" class="statusbar-item left last-visible-item" aria-label="No Problems" title="No Problems"><a tabindex="-1"><span class="codicon codicon-error"></span> 0 <span class="codicon codicon-warning"></span> 0</a></div>
```

**Comment by joanmarie** [comment]
@isidorn: Exactly. That's on the `div`. But Orca is finding a child `a` which is exposed as an accessible link.

**Comment by isidorn** [comment]
@joanmarie aha. Sure we can set the `aria-label` also on `a`, however should we maybe only set the `aria-label` on `a` and not on the parent? I am looking now into status bar and focus and one of the options is to make the `a` elements focusable, not the divs.
What do you think would be the best solution from Orca's standpoint?

**Comment by joanmarie** [comment]
I'm *pretty* sure that Orca will be fine either way -- at least as far as flat review is concerned.

If you put `aria-label` on the `a`, Orca will find the link and consider it presentable to the user because it has a name. Under these conditions, the parent `div` will get ignored because it doesn't have other text (i.e. directly displayed by the `div` rather than child elements). This will be the case regardless of whether or not the `div` has an `aria-label` value.

I don't know if it makes a difference for other screen readers.

**Comment by isidorn** [comment]
I have pushed a change that we only set `aria-label` on the `a` elements.
As for navigating focus I have created this item, so we can think about it then https://github.com/microsoft/vscode/issues/95433

**Comment by jvesouza** [comment]
@joanmarie  It looks like this bug is back. Some items are not read when I use flat review mode to read the contents of the status bar.

The speaks status bar command gives me the following:

'Open a remote window eita (Git) - master* No Problems Select a connection No Notifications Tweet Feedback Select Language Mode Select End of Line Sequence Select Encoding Select Indentation Go to Line/Column If you are not using a Screen Reader, please change the setting `editor.accessibilitySupport` to "off".'{'

The speaks the current flat review line command gives me the following:

'   master*   0    0   Connect Screen Reader Optimized Ln 1, Col 11 Spaces: 4 UTF-8 LF plain Text  '

Am I doing something wrong?

**Comment by isidorn** [comment]
@jvesouza thanks for reporting this. 
@joanmarie the issue is probably that now I only set the `aria-label` on the child `a` elements. Let me know if I should set the `aria-label` on both elements though that feels a bit wrong.

**Comment by joanmarie** [comment]
@isidorn Actually the issue seems to be that Orca is preferring the text of the link rather than the accessible name. More often than not, when a link has text it's name coincides. But that's not necessarily the case. Anyhoo, this is an Orca issue. Will see what I can do. Thanks!

**Comment by joanmarie** [comment]
@jvesouza: Please pull Orca master and re-test. Thanks!

**Comment by jvesouza** [comment]
@joanmarie There are still some small differences in the presentation, but to be honest, that's not a problem, at least for me. The main objective of this issue was to be able to click on the items present in the status bar, and this was already possible without difficulty.

It was great for me. Thank you all.
